File: client/lib/weather_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'url_config_service.dart';
import './utils/greeting_utils.dart';

class WeatherService {
  static WeatherData? _cachedWeatherData;
  static DateTime? _lastFetched;
  static const Duration _cacheDuration = Duration(minutes: 10);

  // Get cached weather data if available and not expired
  static WeatherData? get cachedWeather {
    if (_cachedWeatherData != null && _lastFetched != null) {
      if (DateTime.now().difference(_lastFetched!) < _cacheDuration) {
        return _cachedWeatherData;
      }
    }
    return null;
  }

  // Fetch current weather from the server
  static Future<WeatherData?> getCurrentWeather() async {
    // Return cached data if available
    final cached = cachedWeather;
    if (cached != null) {
      return cached;
    }

    final serverUrl = await UrlConfigService.getServerUrl();
    if (serverUrl == null || serverUrl.isEmpty) {
      return null; // No server configured
    }

    try {
      final response = await http.get(
        Uri.parse('$serverUrl/api/weather/current'),
      ).timeout(const Duration(seconds: 10));

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        final weatherData = WeatherData.fromJson(data);
        
        // Cache the data
        _cachedWeatherData = weatherData;
        _lastFetched = DateTime.now();
        
        return weatherData;
      } else if (response.statusCode == 400) {
        // Weather not configured or disabled
        return null;
      } else {
        throw Exception('Weather API returned status ${response.statusCode}');
      }
    } catch (e) {
      print('Error fetching weather: $e');
      return null; // Return null on error to gracefully degrade
    }
  }

  // Clear cached weather data (useful when configuration changes)
  static void clearCache() {
    _cachedWeatherData = null;
    _lastFetched = null;
  }
}

class WeatherData {
  final CurrentWeather current;
  final LocationInfo location;

  WeatherData({
    required this.current,
    required this.location,
  });

  factory WeatherData.fromJson(Map<String, dynamic> json) {
    return WeatherData(
      current: CurrentWeather.fromJson(json['current'] ?? {}),
      location: LocationInfo.fromJson(json['location'] ?? {}),
    );
  }
}

class CurrentWeather {
  final double temperature;
  final int weatherCode;
  final String weatherIcon;
  final String weatherDesc;
  final double humidity;
  final double windSpeed;
  final double windDirection;
  final double pressure;
  final double visibility;
  final double uvIndex;
  final bool isDay;
  final DateTime lastUpdated;

  CurrentWeather({
    required this.temperature,
    required this.weatherCode,
    required this.weatherIcon,
    required this.weatherDesc,
    required this.humidity,
    required this.windSpeed,
    required this.windDirection,
    required this.pressure,
    required this.visibility,
    required this.uvIndex,
    required this.isDay,
    required this.lastUpdated,
  });

  factory CurrentWeather.fromJson(Map<String, dynamic> json) {
    return CurrentWeather(
      temperature: (json['temperature'] ?? 0.0).toDouble(),
      weatherCode: json['weather_code'] ?? 0,
      weatherIcon: json['weather_icon'] ?? 'unknown',
      weatherDesc: json['weather_desc'] ?? 'Unknown',
      humidity: (json['humidity'] ?? 0.0).toDouble(),
      windSpeed: (json['wind_speed'] ?? 0.0).toDouble(),
      windDirection: (json['wind_direction'] ?? 0.0).toDouble(),
      pressure: (json['pressure'] ?? 0.0).toDouble(),
      visibility: (json['visibility'] ?? 0.0).toDouble(),
      uvIndex: (json['uv_index'] ?? 0.0).toDouble(),
      isDay: json['is_day'] ?? true,
      lastUpdated: DateTime.tryParse(json['last_updated'] ?? '') ?? DateTime.now(),
    );
  }

  // Get temperature with unit
  String get temperatureString => '${temperature.round()}°C';
  
  // Get wind speed with unit
  String get windSpeedString => '${windSpeed.round()} km/h';
  
  // Get weather emoji based on weather code and day/night
  String get weatherEmoji {
    switch (weatherCode) {
      case 0: return isDay ? '☀️' : '🌙'; // Clear
      case 1: case 2: case 3: return isDay ? '⛅' : '☁️'; // Cloudy
      case 45: case 48: return '🌫️'; // Fog
      case 51: case 53: case 55: case 56: case 57: return '🌦️'; // Drizzle
      case 61: case 63: case 65: case 66: case 67: return '🌧️'; // Rain
      case 71: case 73: case 75: case 77: case 85: case 86: return '🌨️'; // Snow
      case 80: case 81: case 82: return '🌦️'; // Rain showers
      case 95: case 96: case 99: return '⛈️'; // Thunderstorm
      default: return isDay ? '☀️' : '🌙';
    }
  }
}

class LocationInfo {
  final double latitude;
  final double longitude;
  final String location;
  final String timezone;
  final DateTime currentTime;

  LocationInfo({
    required this.latitude,
    required this.longitude,
    required this.location,
    required this.timezone,
    required this.currentTime,
  });

  factory LocationInfo.fromJson(Map<String, dynamic> json) {
    return LocationInfo(
      latitude: (json['latitude'] ?? 0.0).toDouble(),
      longitude: (json['longitude'] ?? 0.0).toDouble(),
      location: json['location'] ?? '',
      timezone: json['timezone'] ?? 'UTC',
      currentTime: DateTime.tryParse(json['current_time'] ?? '') ?? DateTime.now(),
    );
  }

  // Get formatted current time
  String get formattedTime {
    final hour = currentTime.hour;
    final minute = currentTime.minute.toString().padLeft(2, '0');
    return '$hour:$minute';
  }

  // Get formatted date
  String get formattedDate {
    final months = [
      'January', 'February', 'March', 'April', 'May', 'June',
      'July', 'August', 'September', 'October', 'November', 'December'
    ];
    final weekdays = [
      'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'
    ];
    
    final weekday = weekdays[currentTime.weekday - 1];
    final month = months[currentTime.month - 1];
    final day = currentTime.day;
    
    return '$weekday, $month $day';
  }

  // Get time of day greeting
  String get greeting {
    return 'Good ${GreetingUtils.getTimeOfDay(dateTime: currentTime).toLowerCase()}';
  }
}

File: server/services/weather_service.go
package services

import (
	"context"
	"fmt"
	"time"

	"github.com/hectormalot/omgo"
	"github.com/azukaar/sumika/server/config"
	"github.com/azukaar/sumika/server/utils"
)

type WeatherService struct {
	cache *WeatherCache
}

type WeatherCache struct {
	data      *WeatherResponse
	expiry    time.Time
	cacheDuration time.Duration
}

type WeatherResponse struct {
	Current  CurrentWeather  `json:"current"`
	Location LocationInfo    `json:"location"`
}

type CurrentWeather struct {
	Temperature     float64   `json:"temperature"`      // Temperature in Celsius
	WeatherCode     int       `json:"weather_code"`     // WMO weather code
	WeatherIcon     string    `json:"weather_icon"`     // Icon identifier
	WeatherDesc     string    `json:"weather_desc"`     // Weather description
	Humidity        float64   `json:"humidity"`         // Relative humidity %
	WindSpeed       float64   `json:"wind_speed"`       // Wind speed km/h
	WindDirection   float64   `json:"wind_direction"`   // Wind direction degrees
	Pressure        float64   `json:"pressure"`         // Surface pressure hPa
	Visibility      float64   `json:"visibility"`       // Visibility km
	UVIndex         float64   `json:"uv_index"`         // UV index
	IsDay           bool      `json:"is_day"`           // Is daytime
	LastUpdated     time.Time `json:"last_updated"`     // Last update time
}

type LocationInfo struct {
	Latitude    float64   `json:"latitude"`
	Longitude   float64   `json:"longitude"`
	Location    string    `json:"location"`
	Timezone    string    `json:"timezone"`
	CurrentTime time.Time `json:"current_time"`
}


type WeatherMapping struct {
	Icon        string
	Description string
}

var WeatherCodeMapping = map[int]WeatherMapping{
	0:  {Icon: "clear-day", Description: "Clear sky"},
	1:  {Icon: "partly-cloudy-day", Description: "Mainly clear"},
	2:  {Icon: "partly-cloudy-day", Description: "Partly cloudy"},
	3:  {Icon: "cloudy", Description: "Overcast"},
	45: {Icon: "fog", Description: "Fog"},
	48: {Icon: "fog", Description: "Depositing rime fog"},
	51: {Icon: "drizzle", Description: "Light drizzle"},
	53: {Icon: "drizzle", Description: "Moderate drizzle"},
	55: {Icon: "drizzle", Description: "Dense drizzle"},
	56: {Icon: "sleet", Description: "Light freezing drizzle"},
	57: {Icon: "sleet", Description: "Dense freezing drizzle"},
	61: {Icon: "rain", Description: "Slight rain"},
	63: {Icon: "rain", Description: "Moderate rain"},
	65: {Icon: "rain", Description: "Heavy rain"},
	66: {Icon: "sleet", Description: "Light freezing rain"},
	67: {Icon: "sleet", Description: "Heavy freezing rain"},
	71: {Icon: "snow", Description: "Slight snow fall"},
	73: {Icon: "snow", Description: "Moderate snow fall"},
	75: {Icon: "snow", Description: "Heavy snow fall"},
	77: {Icon: "snow", Description: "Snow grains"},
	80: {Icon: "rain", Description: "Slight rain showers"},
	81: {Icon: "rain", Description: "Moderate rain showers"},
	82: {Icon: "rain", Description: "Violent rain showers"},
	85: {Icon: "snow", Description: "Slight snow showers"},
	86: {Icon: "snow", Description: "Heavy snow showers"},
	95: {Icon: "thunderstorm", Description: "Thunderstorm"},
	96: {Icon: "thunderstorm", Description: "Thunderstorm with slight hail"},
	99: {Icon: "thunderstorm", Description: "Thunderstorm with heavy hail"},
}

func NewWeatherService() *WeatherService {
	return &WeatherService{
		cache: &WeatherCache{
			cacheDuration: 10 * time.Minute, // Cache weather data for 10 minutes
		},
	}
}

// GetCurrentWeather retrieves current weather data
func (s *WeatherService) GetCurrentWeather() (*WeatherResponse, error) {
	logContext := utils.NewLogContext("weather").WithOperation("get_current")

	// Get configuration
	cfg := config.GetConfig()
	weatherCfg := &cfg.Weather

	// Check if weather is enabled and configured
	if !weatherCfg.Enabled {
		utils.InfoWithContext("Weather is disabled", logContext)
		return nil, fmt.Errorf("weather is disabled")
	}

	if weatherCfg.Latitude == 0.0 && weatherCfg.Longitude == 0.0 {
		utils.InfoWithContext("Weather location not configured", logContext)
		return nil, fmt.Errorf("weather location not configured")
	}

	// Check cache first
	if s.cache.data != nil && time.Now().Before(s.cache.expiry) {
		utils.InfoWithContext("Returning cached weather data", logContext)
		return s.cache.data, nil
	}

	// Fetch weather data from Open-Meteo using omgo
	weatherData, err := s.fetchWeatherFromOpenMeteo(weatherCfg, cfg.Server.Timezone)
	if err != nil {
		utils.ErrorWithContext("Failed to fetch weather data", logContext, err)
		return nil, fmt.Errorf("failed to fetch weather data: %w", err)
	}

	// Cache the data
	s.cache.data = weatherData
	s.cache.expiry = time.Now().Add(s.cache.cacheDuration)

	utils.InfoWithContext("Weather data fetched and cached successfully", logContext)
	return weatherData, nil
}

// fetchWeatherFromOpenMeteo fetches weather data from Open-Meteo API using omgo
func (s *WeatherService) fetchWeatherFromOpenMeteo(weatherCfg *config.WeatherConfig, timezone string) (*WeatherResponse, error) {
	// Create omgo client
	client, err := omgo.NewClient()
	if err != nil {
		return nil, fmt.Errorf("failed to create weather client: %w", err)
	}

	// Create location
	loc, err := omgo.NewLocation(weatherCfg.Latitude, weatherCfg.Longitude)
	if err != nil {
		return nil, fmt.Errorf("failed to create location: %w", err)
	}

	// Set options for the API call
	opts := &omgo.Options{
		Timezone: timezone,
	}

	// Fetch current weather
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	resp, err := client.CurrentWeather(ctx, loc, opts)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch weather: %w", err)
	}

	// Convert omgo response to our weather response format
	weatherResp := s.convertOmgoResponse(&resp, weatherCfg, timezone)
	return weatherResp, nil
}

// convertOmgoResponse converts omgo response to our weather response format
func (s *WeatherService) convertOmgoResponse(resp *omgo.CurrentWeather, weatherCfg *config.WeatherConfig, timezone string) *WeatherResponse {
	// Get weather code value
	weatherCode := int(resp.WeatherCode)

	// Get weather description and icon from weather code
	weatherMapping, exists := WeatherCodeMapping[weatherCode]
	if !exists {
		weatherMapping = WeatherMapping{
			Icon:        "unknown",
			Description: "Unknown",
		}
	}

	// Assume it's day for now (omgo doesn't provide is_day info)
	isDay := true
	icon := weatherMapping.Icon

	// Parse last updated time
	lastUpdated := resp.Time.Time

	// Get current time in the location's timezone
	loc, err := time.LoadLocation(timezone)
	var currentTime time.Time
	if err != nil {
		// Fallback to UTC if timezone is invalid
		currentTime = time.Now().UTC()
	} else {
		currentTime = time.Now().In(loc)
	}

	return &WeatherResponse{
		Current: CurrentWeather{
			Temperature:     resp.Temperature,
			WeatherCode:     weatherCode,
			WeatherIcon:     icon,
			WeatherDesc:     weatherMapping.Description,
			Humidity:        0, // Not available in omgo response
			WindSpeed:       resp.WindSpeed,
			WindDirection:   resp.WindDirection,
			Pressure:        0, // Not available in omgo response
			Visibility:      0, // Not available in omgo response
			UVIndex:         0, // Not available in omgo response
			IsDay:           isDay,
			LastUpdated:     lastUpdated,
		},
		Location: LocationInfo{
			Latitude:    weatherCfg.Latitude,
			Longitude:   weatherCfg.Longitude,
			Location:    weatherCfg.Location,
			Timezone:    timezone,
			CurrentTime: currentTime,
		},
	}
}